File: src/main/java/galaxyoyo/alice/gameengine/editor/CollidPanel.java
package galaxyoyo.alice.gameengine.editor;

import galaxyoyo.alice.gameengine.api.editor.Case;
import galaxyoyo.alice.gameengine.api.editor.Collision;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;

import javax.swing.JPanel;

public class CollidPanel extends JPanel
{
	private static final long serialVersionUID = -138754019431984881L;
	
	private final EditorFrame frame;
	
	public CollidPanel(EditorFrame frame)
	{
		super ();
		this.frame = frame;
	}
	
	public EditorFrame getFrame()
	{
		return frame;
	}
	
	public Map getMap()
	{
		return frame.getMap();
	}
	
	@Override
	public void paintComponent(Graphics g)
	{
		g.fillRect(0, 0, getWidth(), getHeight());
		BufferedImage img = getMap().getFont();
		int x = getWidth() / 2 - img.getWidth();
		int y = getHeight() / 2 - img.getHeight();
		int width = img.getWidth() * 2;
		int height = img.getHeight() * 2;
		g.drawImage(getMap().getFont(), x, y, width, height, null);
		
		for (Case c : getMap().getAllCases())
		{			
			if (isEmpty(c.getCoucheOne().getImage()))
				continue;
			
			g.drawImage(c.getCoucheOne().getImage(), x + c.getPosX() * 34 + 2, y + c.getPosY() * 34 + 2, 32, 32, null);
			
			if (isEmpty(c.getCoucheTwo().getImage()))
				continue;
			
			g.drawImage(c.getCoucheTwo().getImage(), x + c.getPosX() * 34 + 2, y + c.getPosY() * 34 + 2, 32, 32, null);
			
			if (isEmpty(c.getCoucheThree().getImage()))
				continue;
			
			g.drawImage(c.getCoucheThree().getImage(), x + c.getPosX() * 34 + 2, y + c.getPosY() * 34 + 2, 32, 32, null);
		}
		
		for (Case c : getMap().getAllCases())
		{
			if (c.getCollision() != Collision.ANY)
			{
				BufferedImage alpha = new BufferedImage(16, 16, BufferedImage.TYPE_INT_ARGB);
				
				if (c.getCollision() == Collision.FULL)
				{
					Graphics2D grap = alpha.createGraphics();
					grap.setColor(new Color(0, 0, 0, 100));
					grap.fillRect(0, 0, 16, 16);
					grap.dispose();
				}
				else if (c.getCollision() == Collision.PARTIAL)
				{
					Graphics2D grap = alpha.createGraphics();
					grap.setColor(new Color(255, 0, 255, 70));
					grap.fillRect(0, 0, 16, 16);
					grap.dispose();
				}
				
				g.drawImage(alpha, x + c.getPosX() * 34 + 2, y + c.getPosY() * 34 + 2, 32, 32, null);
			}
		}
	}

	private boolean isEmpty(BufferedImage image)
	{
		int allrgba = 0;
		
		for (int x = 0; x < image.getWidth(); ++x)
		{
			for (int y = 0; y < image.getHeight(); ++y)
			{
				allrgba += image.getRGB(x, y) + 1;
			}
		}
		
		return allrgba == 0;
	}
}

File: src/main/scala/fr/ynerant/leveleditor/editor/CollidPanel.scala
package fr.ynerant.leveleditor.editor

import java.awt._
import java.awt.image.BufferedImage

import fr.ynerant.leveleditor.api.editor.Collision
import javax.swing._


@SerialVersionUID(-138754019431984881L)
object CollidPanel {
	def isEmpty(image: BufferedImage): Boolean = {
		var allrgba = 0
		for (x <- 0 until image.getWidth) {
			for (y <- 0 until image.getHeight) {
				allrgba += image.getRGB(x, y) + 1
			}
		}
		allrgba == 0
	}
}

@SerialVersionUID(-138754019431984881L)
class CollidPanel(val frame: EditorFrame) extends JPanel {
	override def paintComponent(g: Graphics): Unit = {
		g.fillRect(0, 0, getWidth, getHeight)
		val img = getMap.getFont
		val x = getWidth / 2 - img.getWidth
		val y = getHeight / 2 - img.getHeight
		val width = img.getWidth * 2
		val height = img.getHeight * 2
		g.drawImage(getMap.getFont, x, y, width, height, null)
		getMap.getAllCases.foreach(c => {
			if (!CollidPanel.isEmpty(c.getCoucheOne.getImage)) {
				g.drawImage(c.getCoucheOne.getImage, x + c.getPosX * 34 + 2, y + c.getPosY * 34 + 2, 32, 32, null)
				if (!CollidPanel.isEmpty(c.getCoucheTwo.getImage)) {
					g.drawImage(c.getCoucheTwo.getImage, x + c.getPosX * 34 + 2, y + c.getPosY * 34 + 2, 32, 32, null)
					if (!CollidPanel.isEmpty(c.getCoucheThree.getImage))
						g.drawImage(c.getCoucheThree.getImage, x + c.getPosX * 34 + 2, y + c.getPosY * 34 + 2, 32, 32, null)
				}
			}
		})

		getMap.getAllCases.foreach(c => {
			if (c.getCollision.equals(Collision.ANY)) {
				val alpha = new BufferedImage(16, 16, BufferedImage.TYPE_INT_ARGB)
				if (c.getCollision eq Collision.FULL) {
					val grap = alpha.createGraphics
					grap.setColor(new Color(0, 0, 0, 100))
					grap.fillRect(0, 0, 16, 16)
					grap.dispose()
				}
				else if (c.getCollision eq Collision.PARTIAL) {
					val grap = alpha.createGraphics
					grap.setColor(new Color(255, 0, 255, 70))
					grap.fillRect(0, 0, 16, 16)
					grap.dispose()
				}
				g.drawImage(alpha, x + c.getPosX * 34 + 2, y + c.getPosY * 34 + 2, 32, 32, null)
			}
		})
	}

	def getMap: GMap = frame.getMap
}
